### PR TITLE
pull in changes from droplet master (including droplet svg) and fix regressions

### PR DESCRIPTION
**Pull request by davidsbailey**
Sync our copy of droplet to https://github.com/droplet-editor/droplet/commit/c659d021d2da9662c1ff4e757747c61674c3e117 and fix regressions.

The biggest change here is that droplet has been reimplemented in SVG instead of Canvas. There are also a few follow-on fixes related to that, most notably reimplementing horizontal scrolling in https://github.com/droplet-editor/droplet/commit/a6918aeafa4019f43fffeff8770f821887435b6d.

Notable changes not related to SVG:
* https://github.com/droplet-editor/droplet/commit/e6ea0715b56bf5c5f4126e98fe83b75b14ca5dce add,subtract buttons on array expression
* https://github.com/droplet-editor/droplet/commit/63bfd0b8fe5a5142281c3c8187d0a2260cf8f1a9 remove expansion properties when dropping block from palette into doc…
* https://github.com/droplet-editor/droplet/commit/5f7bece03f1ee9511ba60bdcdf881627d94f405e don't allow dropping inside sockets that are in comments
* https://github.com/droplet-editor/droplet/commit/c7f68f60455ad04a6c9801ea9c666eb8066913af add/remove params from FunctionDeclaration
* https://github.com/droplet-editor/droplet/pull/200 implement subtract button on else clauses

### CODE REVIEW

**Comment by davidsbailey** [comment]
I've updated https://dashboard-adhoc-assigncoursefromtd.cdn-code.org/projects/applab/ with the contents of this PR in case you want to try it out.

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -1055,7 +1055,7 @@ StudioApp.prototype.showInstructionsDialog_ = function (level, autoClose) {
 
   var hideFn = _.bind(function () {
     // Set focus to ace editor when instructions close:
-    if (this.editCode && this.editor && !this.editor.currentlyUsingBlocks) {
+    if (this.editCode && this.editor && !this.editor.session.currentlyUsingBlocks) {
```
Comment (cpirich) [inline apps/src/StudioApp.js]: Do we need to check for `editor.session` first here? (like you do in `resizeToolbarHeader()`?)

File: apps/src/StudioApp.js [hunk 2]
```diff
@@ -2117,7 +2127,7 @@ StudioApp.prototype.handleEditCode_ = function (config) {
       if (range) {
         var lineIndex = range.start.row;
         var line = lineIndex + 1; // 1-based line number
-        if (this.editor.currentlyUsingBlocks) {
+        if (this.editor.session.currentlyUsingBlocks) {
```
Comment (cpirich) [inline apps/src/StudioApp.js]: check for `session`?

File: apps/src/StudioApp.js [hunk 3]
```diff
@@ -2325,7 +2335,7 @@ StudioApp.prototype.handleUsingBlockly_ = function (config) {
  */
 StudioApp.prototype.onDropletToggle = function (autoFocus) {
   autoFocus = utils.valueOr(autoFocus, true);
-  if (!this.editor.currentlyUsingBlocks) {
+  if (!this.editor.session.currentlyUsingBlocks) {
```
Comment (cpirich) [inline apps/src/StudioApp.js]: check for `session`?

File: apps/src/templates/ShowCodeToggle.js [hunk 4]
```diff
@@ -99,7 +99,7 @@ class DropletCodeToggle extends Component {
 
   toggle = () => {
     // are we trying to toggle from blocks to text (or the opposite)
-    let fromBlocks = studioApp().editor.currentlyUsingBlocks;
+    let fromBlocks = studioApp().editor.session.currentlyUsingBlocks;
```
Comment (cpirich) [inline apps/src/templates/ShowCodeToggle.js]: check for `session`?

File: apps/src/templates/ShowCodeToggle.js [hunk 5]
```diff
@@ -116,15 +116,15 @@ class DropletCodeToggle extends Component {
     } else {
       studioApp().onDropletToggle();
       this.forceUpdate();
-      this.props.onToggle(studioApp().editor.currentlyUsingBlocks);
+      this.props.onToggle(studioApp().editor.session.currentlyUsingBlocks);
```
Comment (cpirich) [inline apps/src/templates/ShowCodeToggle.js]: check for `session`?

File: apps/src/templates/ShowCodeToggle.js [hunk 6]
```diff
@@ -116,15 +116,15 @@ class DropletCodeToggle extends Component {
...
     }
   }
 
   onClick = () => {
     this.toggle();
     if (experiments.isEnabled('saveBlockMode')) {
       new UserPreferences().setUsingTextMode(
-        !studioApp().editor.currentlyUsingBlocks,
+        !studioApp().editor.session.currentlyUsingBlocks,
```
Comment (cpirich) [inline apps/src/templates/ShowCodeToggle.js]: check for `session`?

**Comment by davidsbailey** [comment]
Inspecting the droplet code reveals that editor.session could be null, so I've extracted this into a helper method.

**Comment by davidsbailey** [comment]
@cpirich I've updated this PR to include the subtract button on else clauses. PTAL

**Pull request merged**